plugins/mqtt/mqtt_util.c

```diff
@@ -0,0 +1,249 @@
+        return -1;
+    }
+
+    int             ret       = 0;
+    char *          error     = NULL;
+    neu_json_elem_t id        = { .name = "client-id", .t = NEU_JSON_STR };
+    neu_json_elem_t upload    = { .name = "upload-topic", .t = NEU_JSON_STR };
+    neu_json_elem_t heartbeat = { .name = "heartbeat-topic",
+                                  .t    = NEU_JSON_STR };
+    neu_json_elem_t format    = { .name = "format", .t = NEU_JSON_INT };
+    neu_json_elem_t ssl       = { .name = "ssl", .t = NEU_JSON_BOOL };
+    neu_json_elem_t host      = { .name = "host", .t = NEU_JSON_STR };
+    neu_json_elem_t port      = { .name = "port", .t = NEU_JSON_INT };
+    neu_json_elem_t username  = { .name = "username", .t = NEU_JSON_STR };
+    neu_json_elem_t password  = { .name = "password", .t = NEU_JSON_STR };
+    neu_json_elem_t ca        = { .name = "ca", .t = NEU_JSON_STR };
+    neu_json_elem_t cert      = { .name = "cert", .t = NEU_JSON_STR };
+    neu_json_elem_t key       = { .name = "key", .t = NEU_JSON_STR };
+    neu_json_elem_t keypass   = { .name = "keypass", .t = NEU_JSON_STR };
+
+    // client-id, required
+    ret = neu_parse_param(config, &error, 1, &id);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode client-id from json");
+        free(id.v.val_str);
+        return -1;
+    } else {
+        option->clientid = id.v.val_str;
+    }
+
+    // upload-topic, optional
+    ret = neu_parse_param(config, &error, 1, &upload);
+    if (0 != ret) {
+        free(upload.v.val_str);
+    } else {
+        option->upload_topic = upload.v.val_str;
+    }
+
+    // heartbeat-topic, optional
+    ret = neu_parse_param(config, &error, 1, &heartbeat);
+    if (0 != ret) {
+        free(heartbeat.v.val_str);
+    } else {
+        option->heartbeat_topic = heartbeat.v.val_str;
+    }
+
+    // format, required
+    ret = neu_parse_param(config, &error, 1, &format);
+    if (0 != ret) {
+        option->format = 0;
+    } else {
+        option->format = format.v.val_int;
+    }
+
+    // ssl, required
+    ret = neu_parse_param(config, &error, 1, &ssl);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode ssl from json");
+        return -2;
+    } else {
+        if (false == ssl.v.val_bool) {
+            option->connection = strdup("tcp://");
+        } else {
+            option->connection = strdup("ssl://");
+
+            // ca, required if ssl enabled
+            ret = neu_parse_param(config, &error, 1, &ca);
+            if (0 != ret) {
+                plog_error(plugin, "can't decode ca from json");
+                free(ca.v.val_str);
+                return -3;
+            } else {
+                option->ca = ca.v.val_str;
+            }
+
+            // cert, optional
+            ret = neu_parse_param(config, &error, 1, &cert);
+            if (0 == ret) {
+                free(cert.v.val_str);
+            } else {
+                option->cert = cert.v.val_str;
```

Comment:
relatedd fields can be parsed at once

Reply:
It would be better to keep the status quo here.